docs: ADR-008 to restructure the `ibc` crate (#966)
cosmos/ibc-rs: docs/architecture/adr-008-restructure-ibc-crate.md

This ADR aims to enhance both the usability and practicality of `ibc-rs` by restructuring the codebase and organizing it under multiple sub-libraries, as stated in the [decision](#decision) section. This will make different parts of `ibc-rs` accessible to users, positioning it as a more comprehensive, one-stop solution catering to diverse users groups, whether for on-chain or off-chain use cases.

## Decision

For the library organization, the first stage of separation is to split off the codebase of each IBC applications, clients, and core implementation, decoupling them from each other. The top-level libraries and the naming schema would look like as follow:

```markdown
.
├── ibc -> Primarily re-exports sub-libraries
├── ibc-core 
│   ├── ibc-core-client (contains the implementation + Re-exports types)
│   ├── ibc-core-connection
│   ├── ibc-core-channel
│   ├── ibc-core-commitment
│   └── ibc-core-host
│       └── .
├── ibc-clients
│   ├── ibc-client-tendermint

> │   ├── ibc-client-tendermint-cw

Reviewer: Should this be `ibc-client-cosmwasm`?

Author: I was in favor of setting up a whole new crate for Wasm clients at first. Though, in practice I am seeing that's just extra crate we don't need. We can keep it simple and throw all the light clients implementation under the `ibc-clients`. And to tell apart the CosmWasm implementation of a specific light client, let's just append a `-cw` at the end. Therefore for `ibc-client-tendermint` the CW implementation would be maintained under `ibc-client-tendermint-cw`

      implementation is impractical.
2. **Selective Types Import**
    - Relayers, like Hermes, or any off-chain consumers cannot import their
      desired layer of implementation like ibc types without pulling in
      unnecessary dependencies into their project.
3. **Smoother IBC Core Integration with Hosts**
    - Integrating ibc core with host chains without introducing light client or
      app dependencies is currently not straightforward, impeding smooth
      integration.
4. **Easier Development of CosmWasm Contracts**
    - For developing a CosmWasm tendermint light client, we ideally should only
      be dependent on implementation under the **`ics07_tendermint`** and also
      be importing relevant parts from the **`ibc-core-client`** layer without
      pulling in all the ibc codebase and dependencies.

This ADR aims to enhance both the usability and practicality of `ibc-rs` by
restructuring the codebase and organizing it under multiple sub-libraries, as
stated in the [decision](#decision) section. This will make different parts of
`ibc-rs` accessible to users, positioning it as a more comprehensive, one-stop

> solution catering to diverse users groups, whether for on-chain or off-chain use

Reviewer: ```suggestion
solution catering to diverse user groups, whether for on-chain or off-chain use
```

cases.

## Decision

For the library organization, the first stage of separation is to split off the
codebase of each IBC applications, clients, and core implementation, decoupling
them from each other. The top-level libraries and the naming schema would look

> like as follow:

Reviewer: ```suggestion
For the library organization, the first stage of separation is to split the
codebase so that each IBC application, client, and core implementation is decoupled from one another. The top-level libraries and the naming schema would look as follows:
```

```markdown
.
├── ibc -> Primarily re-exports sub-libraries
├── ibc-core 
│   ├── ibc-core-client (contains the implementation + Re-exports types)
│   ├── ibc-core-connection
│   ├── ibc-core-channel
│   ├── ibc-core-commitment
│   └── ibc-core-host
│       └── .
├── ibc-clients
│   ├── ibc-client-tendermint
│   ├── ibc-client-tendermint-cw
│   └── .
├── ibc-apps
│   ├── ibc-app-transfer
│   ├── ibc-app-ica
│   │   └── .
│   └── .
├── ibc-preludes
├── ibc-testkit (previously mock module + `test-utils` feature)
├── ibc-query
└── ibc-derive
```

With this restructure, the main `ibc` crate primarily re-exports types,

> interfaces, and implementation of all the sub-libraries. Therefore, if someone

Reviewer: ```suggestion
interfaces, and implementations of all the sub-libraries. Therefore, if someone
```

only wants to depend on the `ibc` crate without caring about this granularity,
they can do so.

> Afterward, we split off data structure of each IBC layer into a separate

Reviewer: ```suggestion
Afterward, we split off the data structures of each IBC layer into a separate
```

sub-library under a `types` folder, still maintained under the directory of that
relevant component/module. As an example, the `ibc-core-client` crate’s tree and
the naming schema looks like this:

```markdown
ibc-core
└── ibc-core-client (dir: ibc-core/ics02-client)
    └── ibc-core-client-types (dir: ibc-core/ics02-client/types)
        ├── msgs
        ├── events
        └── .
```

This way, the main crate of each IBC module contains all the necessary APIs and

> implementation to integrate with host chains, along with re-exporting the

Reviewer: ```suggestion
implementations to integrate with host chains, along with re-exporting the
```

sub-library types. This allows projects to selectively import types (e.g.
`ibc-core-client-types`), often required by off-chain users such as relayers. Or
to pick the library containing the entire implementation of that particular
module (e.g. `ibc-core-client`), typically more convenient for host chains or

> smart contract developers to integrate with their end.

Reviewer: ```suggestion
smart contract developers to integrate with on their end.
```

By this restructuring, the **directory tree** of the repo would look like as

> follow:

Reviewer: ```suggestion
Once the restructuring is complete, the **directory tree** of the repo would look as follows:
```

```markdown
ibc
ibc-core
├── ics02-client
|   ├── src
|   ├── types
|   |   ├── src
|   |   └── cargo.toml
|   └── cargo.toml
├── ics03-connection
|   └── .
├── ics04-channel
|   └── .
├── ics23-commitment
|   └── .
├── ics24-host
|   └── .
├── src

> ├── cargo.toml

Reviewer: ```suggestion
├── Cargo.toml
```

└── README.md
ibc-clients
├── ics07-tendermint
├── ics08-wasm
└── .
ibc-apps
├── ics20-transfer
├── ics27-ica
└── .
ibc-preludes
ibc-testkit
ibc-query
ibc-derive
```

To implement this change efficiently and for more organization, we use the
workspace inheritance feature and will add a top-level README for main library
groups like `ibc-core`, `ibc-clients`, etc serving as a guide for users to
understand the purpose and structure of their sub libraries.

Later, to streamline our release process, it is crucial to come up with a Github
action to automate and simplify the release process.

## **Status**

Proposed

## **Consequences**

We should acknowledge this restructuring, while a significant step forward, will
not completely address all existing design couplings. Subsequent improvements in
implementation logic will be necessary to completely decouple ibc core, clients,
and applications from each other and make the entire logic as chain-agnostic as

> possible. For an instance, currently, our `IbcEvent` type depends on the

Reviewer: ```suggestion
possible. 
For instance, currently, our `IbcEvent` type depends on the
```

> Tendermint events in their conversion, which needs to be addressed afterward.

Reviewer: ```suggestion
Tendermint events in their conversion, which can only be addressed once this restructuring is complete.
```
